### ISSUE TITLE
lsp-mode does not work with latest un-merged but published version of eslint language server

### ISSUE THREAD

**Issue by artemkovalyov** [emacs-lsp/lsp-mode#1932]
All the tests were done with clean emacs 27+ and config from here: `https://github.com/emacs-lsp/lsp-mode/blob/master/scripts/lsp-start-plain.el`.

**Describe the bug**
Eslint doesn't apply fixes. `lsp-eslint-apply-all-fixes` doesn't produce any results.

**To Reproduce**
Create a following `test.js` file.
```js
function abc ()         {
  return 23;;
}

abc(

)
```
Create `package.json` file:
```json
{
  "name": "lsp",
  "version": "1.0.0",
  "description": "",
  "main": "test.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "standard": "^14.3.4"
  }
}
```
Create `.eslintrc.js`
```js
module.exports = {
  env: {
    browser: true,
    es6: true
  },
  extends: [
    'standard'
  ]
}
```

**Run**

`npm install`

**Command line tools work!**
```bash
/data/tmp/lsp 
❯ eslint test.js
/data/tmp/lsp/test.js
1:16  error  Multiple spaces found before '{'    no-multi-spaces
2:12  error  Extra semicolon                     semi
2:12  error  Missing whitespace after semicolon  semi-spacing
✖ 3 problems (3 errors, 0 warnings)
3 errors and 0 warnings potentially fixable with the `--fix` option.
```
When --fix option is applied it fixes all the errors in the file.

**Expected behavior**
1. Errors are highlighted.
2. `lsp-eslint-apply-all-fixes` fixes what is automatically possible.

**Which Language Server did you use**
dbaeumer.vscode-eslint-2.1.8.vsix with `lsp-eslint`

**OS**
Linux

** LSP Log**
```
[Trace - 09:59:07 PM] Received response 'textDocument/codeAction - (202)' in 17ms.
Result: []


[Trace - 09:59:14 PM] Sending request 'workspace/executeCommand - (204)'.
Params: {
  "command": "eslint.applyAllFixes",
  "arguments": [
    {
      "uri": "file:///data/dev/cloud-sdk/src/test.js",
      "version": 4
    }
  ]
}


[Trace - 09:59:14 PM] Received request 'workspace/applyEdit - (9).
Params: {
  "edit": {
    "documentChanges": [
      {
        "edits": [],
        "textDocument": {
          "version": 4,
          "uri": "file:///data/dev/cloud-sdk/src/test.js"
        }
      }
    ]
  }
}


[Trace - 09:59:14 PM] Sending response 'workspace/applyEdit - (9)'. Processing request took 2ms
Params: {
  "jsonrpc": "2.0",
  "id": 9,
  "result": {
    "applied": true
  }
}


[Trace - 09:59:14 PM] Received response 'workspace/executeCommand - (204)' in 17ms.
Result: {
}
```

**Comment by yyoncho** [comment]
AFACS the server code has changed so we have to adjust lsp-mode.

**Comment by artemkovalyov** [comment]
A couple of screenshots to get an idea.
![image](https://user-images.githubusercontent.com/10460752/87829124-ab15f980-c87e-11ea-962c-e8e108621c34.png)
![image](https://user-images.githubusercontent.com/10460752/87829141-b1a47100-c87e-11ea-9e1c-3f95670330ea.png)

**Comment by artemkovalyov** [comment]
Understood @yyoncho, at least I won't be duping time to configure it in vain and make a gig to make it work again.

**Comment by yyoncho** [comment]
Yes. The issue is that the server hasn't been configured properly before that so it is sending empty edits when performing quick fix.

Probably if you try to switch to an older version of the server it will work. The issue might be caused also by a regression in lsp-mode as well. I will take a look to find out how much effort will be to sync lsp-mode with latest eslint.

**Comment by yyoncho** [comment]
I downloaded https://github.com/Microsoft/vscode-eslint and then npm install and then configured lsp-mode to point to vscode-eslint/server/out/eslintServer.js and it was working fine.

At the same time, it does not work with the eslint downloaded from market place. I cannot find which version was used for 2.1.8 version of vscode package. Also, I got a notification request when I was using the marketplace version('eslint/confirmLocalESLint') but I was unable to find that in the github sources so it seems to me that this was not released from GH repo.

**Comment by leungbk** [comment]
The commits corresponding to the latest versions haven't been merged into the master branch yet: https://github.com/microsoft/vscode-eslint/issues/1011

**Comment by yyoncho** [comment]
> The commits corresponding to the latest versions haven't been merged into the master branch yet: [microsoft/vscode-eslint#1011](https://github.com/microsoft/vscode-eslint/issues/1011)

This is really something unexpected...

**Comment by artemkovalyov** [comment]
It's indeed weird and unexpected... Thanks for looking into this @yyoncho and @leungbk

**Comment by leungbk** [comment]
So this is a bit messy. The new `esilnt/confirmLocalESLint` request introduced in https://github.com/microsoft/vscode-eslint/commit/455cab65ec8b687d20d1efa53828a93110c1f2fd on the `2.1.9` branch is the cause of the problem.

Upstream, they are still debating on the best way of doing this: https://github.com/microsoft/vscode-eslint/issues/1012. I don't think we can work around this by merely sending `Allow` every time since `eslint` is a mult-root server.

**Comment by Rogergonzalez21** [comment]
For anyone that still has this problem, I reverted back to [2.1.5](https://github.com/microsoft/vscode-eslint/releases/tag/release%2F2.1.5), `npm install`ed it, `npm run compile`d it, changed the `lsp-eslint-server-command` location and it works. This is a good workaround until this issue gets resolved :+1:

**Comment by danielpza** [comment]
You can also add a dummy handler for the request to ignore it. Although I think we should manage this request properly at some point:

```diff
modified   clients/lsp-eslint.el
@@ -210,6 +210,10 @@ source.fixAll code action."
   "Open documentation."
   (browse-url url))
 
+(lsp-defun lsp-eslint--confirm-local-eslint ()
+  "Confirm Local Eslint"
+  nil) ;; TODO
+
 (defun lsp-eslint-apply-all-fixes ()
   "Apply all autofixes in the current buffer."
   (interactive)
@@ -248,7 +252,8 @@ source.fixAll code action."
   :multi-root t
   :notification-handlers (ht ("eslint/status" #'lsp-eslint-status-handler))
   :request-handlers (ht ("workspace/configuration" #'lsp-eslint--configuration)
-                        ("eslint/openDoc" 'lsp-eslint--open-doc))
+                        ("eslint/openDoc" 'lsp-eslint--open-doc)
+                        ("eslint/confirmLocalEslint" 'lsp-eslint--confirm-local-eslint))
   :server-id 'eslint
   :initialized-fn (lambda (workspace)
                     (with-lsp-workspace workspace

```

Other requests not handled? https://github.com/microsoft/vscode-eslint/pull/1053#issuecomment-681760741

**Comment by leungbk** [comment]
> You can also add a dummy handler for the request to ignore it. Although I think we should manage this request properly at some point:
> 
> ```diff
> modified   clients/lsp-eslint.el
> @@ -210,6 +210,10 @@ source.fixAll code action."
>    "Open documentation."
>    (browse-url url))
>  
> +(lsp-defun lsp-eslint--confirm-local-eslint ()
> +  "Confirm Local Eslint"
> +  nil) ;; TODO
> +
>  (defun lsp-eslint-apply-all-fixes ()
>    "Apply all autofixes in the current buffer."
>    (interactive)
> @@ -248,7 +252,8 @@ source.fixAll code action."
>    :multi-root t
>    :notification-handlers (ht ("eslint/status" #'lsp-eslint-status-handler))
>    :request-handlers (ht ("workspace/configuration" #'lsp-eslint--configuration)
> -                        ("eslint/openDoc" 'lsp-eslint--open-doc))
> +                        ("eslint/openDoc" 'lsp-eslint--open-doc)
> +                        ("eslint/confirmLocalEslint" 'lsp-eslint--confirm-local-eslint))
>    :server-id 'eslint
>    :initialized-fn (lambda (workspace)
>                      (with-lsp-workspace workspace
> ```

I thought about doing this a while ago. It seems like they're working toward a solution upstream (https://github.com/microsoft/vscode-eslint/issues/1012#issuecomment-684870074), but it's not clear when it'll be finalized. So I think for now, we should go with this. A PR would be appreciated.

**Comment by yyoncho** [comment]
Reopening - they just published 2.1.0

**Comment by adimit** [comment]
2.1.10 even. I had to revert to vscode-eslint 2.0.9, probably because of how executing [eslint libraries](https://github.com/Microsoft/vscode-eslint#version-2110) has changed: it can't find eslint-prettier (works fine from command line.) So There should be some way of going through the motions described in the vscode-eslint README

**Comment by komali2** [comment]
Is this the thread should be following if I'm having this issue? Whenever I open a JS file i get several of this error

```
    Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
    Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
    Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
    Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
```

I'm using lsp-mode I think through flychecker, I could be wrong as this part of emacs usage is a little complicated to me still. If I do `flycheck-verify-setup` this is my output. 

lsp-mode seems to work for .vue files just fine. 

```

Syntax checkers for buffer media-with-content.spec.js in js2-mode:

First checker to run:

  lsp (explicitly selected)
    - may enable: yes
    - may run:    t

```

**Comment by adimit** [comment]
@komali2 Have you tried it with vscode-eslint version 2.0.9? I had the same problem

```
    Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
```
This is about the eslint libraries/modules, I presume.

**Comment by komali2** [comment]
Oh, huh, uh, these errors are coming to me from emacs. I think my stack looks like lsp-mode talking to my local node_modules eslint, which is version `7.1.0`. I'm very confused, clearly. But in any case it's my impression that vscode isn't involved at all?

**Comment by yyoncho** [comment]
Downgrading 2.1.8 should work.

**Comment by komali2** [comment]
version 2.1.8 of what?

**Comment by yyoncho** [comment]
Of eslint extension - https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint

**Comment by komali2** [comment]
I don't use visual studio code, so I'm not sure how installing a different version of a vscode extension would help me.

**Comment by adimit** [comment]
@komali2 The installation path for the eslint LSP server is either through the VSCode extension, or through the repository at https://github.com/Microsoft/vscode-eslint.

In my experience, though, if you're using eslint plugins like prettier, you'll need to use 2.0.9, not 2.1.8. You could try out either version.

**Comment by drocha87** [comment]
@komali2 the solution proposed here fixed your problem? I'm having the exactly same problem and I didn't understand how to fix those warning messages.

**Comment by zanemayo** [comment]
Installing version 2.1.8 of vscode-eslint worked for me.

I downloaded it from here: https://github.com/microsoft/vscode-eslint/releases/tag/release%2F2.1.8
I unzipped the .vsix file into .emacs.d/.cache/lsp/eslint/unzipped (but first removed the files that were there currently)
Then I started emacs again and it worked.

**Comment by artemkovalyov** [comment]
@zanemayo, I'll give it a go today, thanks for testing.

**Comment by drocha87** [comment]
@zanemayo I did exactly what you said and seems to work just fine. I'll keep testing it during the day and report again later, thanks!

**Comment by komali2** [comment]
@zanemayo 's solution worked for me. Thanks zanemayo for the explicit instructions!

**Comment by yyoncho** [comment]
FTR I forced automatic installation to download and install 2.1.8 version. Thus, you will be able to install eslint by doing M-x lsp-install-server and it will install the right versions. This issue will be still open, but it will be with much lower priority since the default will just work.

**Comment by artemkovalyov** [comment]
@yyoncho, I pullet the latest lsp-mode from GH with straight and I don't see an option to install a eslint server. Should I do some extra steps to make it work?
![image](https://user-images.githubusercontent.com/10460752/99160385-31d0cb00-26e7-11eb-80f3-fc261bb21d65.png)

**Comment by artemkovalyov** [comment]
By manually installing 2.1.8, I can confirm that `lsp-eslint-apply-all-fixes' works. For me, it does the job on the whole buffer. With version 2.1.10 it doesn't work.

**Comment by yyoncho** [comment]
> @yyoncho, I pullet the latest lsp-mode from GH with straight and I don't see an option to install a eslint server. Should I do some extra steps to make it work?

You have to remove the manually installed one and reset all lsp-eslint related settings.

**Comment by elthariel** [comment]
I have the same annoying warning (`Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution
`) on my setup. I've been trying versions 1.9.1, 2.1.8, 2.0.9 of vscode-eslint.

**Comment by arjan** [comment]
2.1.8 works for me

**Comment by elthariel** [comment]
Strangely, using the 2.1.8 version pulled by `install-lsp-server` doesn't display this warning, but using the release/2.1.8 tag from the git repository does (using a clean clone and running  `npm i && npm run webpack`

**Comment by farynaio** [comment]
> Strangely, using the 2.1.8 version pulled by `install-lsp-server` doesn't display this warning, but using the release/2.1.8 tag from the git repository does (using a clean clone and running `npm i && npm run webpack`

How to get specific version with `lsp-install-server`?

**Comment by RKBK** [comment]
I ran into this issue only after updating to lsp-mode 20210120.2145, but following @yyoncho's recommendation to delete the language server (delete the subfolder "eslint" inside ~/.emacs.d/.cache/lsp), restarting emacs, and then lsp-install-server eslint, seems to have fixed it.

**Comment by clayrisser** [comment]
I'm getting the warning `Warning (lsp-mode): Unknown request method: eslint/confirmESLintExecution` with `2.1.8`. When I use the latest pre-release it works with no errors.

https://github.com/microsoft/vscode-eslint/releases/download/2.1.12-Insider/vscode-eslint-2.1.12.vsix

**Comment by nbfalcon** [comment]
Update and recompile `lsp-mode` again, because there is a message handler for that @clayrisser

**Comment by clayrisser** [comment]
And thanks. How do I do that?

**Comment by nbfalcon** [comment]
Do you use a distribution like Spacemacs/Doom or do you have a custom config?

**Comment by nbfalcon** [comment]
E.g. for `doom` that would be `doom/reload`, for Spacemacs `SPC feU` in the status buffer, and if you use `package.el` with a custom config, then [this](https://stackoverflow.com/questions/14836958/updating-packages-in-emacs).
